cluster/kubernetes/manifests.go

```diff
@@ -26,11 +33,39 @@ type Manifests struct {
 	Namespacer namespacer
 }
 
+func getCRDScopes(manifests map[string]kresource.KubeManifest) ResourceDefinitionScopes {
+	result := ResourceDefinitionScopes{}
+	for _, km := range manifests {
+		if km.GetKind() == "CustomResourceDefinition" {
+			var crd v1beta1.CustomResourceDefinition
+			if err := yaml.Unmarshal(km.Bytes(), &crd); err != nil {
+				// The CRD can't be parsed, so we (intentionally) ignore it and
+				// just hope for EffectiveNamespace() to find its scope in the cluster if needed.
+				continue
+			}
+			crdVersions := crd.Spec.Versions
+			if len(crdVersions) == 0 {
+				crdVersions = []v1beta1.CustomResourceDefinitionVersion{{Name: crd.Spec.Version}}
+			}
+			for _, crdVersion := range crdVersions {
+				gvk := schema.GroupVersionKind{
+					Group:   crd.Spec.Group,
+					Version: crdVersion.Name,
+					Kind:    crd.Spec.Names.Kind,
+				}
+				result[gvk] = crd.Spec.Scope
+			}
+		}
+	}
+	return result
+}
+
 func postProcess(manifests map[string]kresource.KubeManifest, nser namespacer) (map[string]resource.Resource, error) {
+	knownScopes := getCRDScopes(manifests)
```

Comment:
Is it possible to use the fallback lookup here instead of putting it within `EffectiveNamespace`? It jars a little to have some of the special case code here (collecting the CRD scopes) and some of it there -- it feels like they should just be in the same place.

Reply:
Uhm, that's how I intended to implement it initially, but I though this would be cleaner. I will give it another go.

Reply:
Oh dear am I too unpredictable as a reviewer :-S

Reply:
Hehehe, no worries. I gave it a try but I don't see a clean way to do it without replicating `namespaceViaDiscovery.EffectiveNamespace` in `manifests.go`.

I concur in that splitting the obtention of the scope - as per the cluster and as per the git manifests - in separate files is ugly. However for the later you need all the resources.

I could merge `namespacer.go` into `manifests.go` and remove the namespacer interface if you think that's better.

Reply:
I'm OK with having that split, so long as it's clean: that is, namespacer.go does what it can to answer with respect to the cluster, and if it cannot, postProcess in manifests.go looks it up in the map of scopes it has just compiled.

Maybe there's a reason it won't work as I imagine it -- I'm fine with how it is, really.

Reply:
Then let's merge it as is. I am ready to push the button after you've taken another look

Reply:
I worked out how I'm being dim.

It's not enough to have a record of which custom resources are cluster scoped and which are not, to answer the question "which namespace should this go in?"  -- you also need to know the fallback namespace. So it can't be done in manifests.go without further changes to make that latter available, and it wouldn't be any simpler.

Reply:
Exactly :)